docs/sessions.md

```diff
@@ -0,0 +1,127 @@
+
+/// WebPart
+let app =
+  statefulForSession
+  >=> setSessionValue "test" "123"
+  >=> choose [
+        path "/yes" >=> cookieYes
+        path "/no" >=> cookieNo
+        RequestErrors.NOT_FOUND
+        ]
+{% endhighlight %}
+
+Server Keys
+-----------
+
+The contents of the cookie are encrypted before the cookie is sent. Suave's default configuration generates a new server key each time the server is restarted. While this is not _wrong_, users would likely get quite annoyed if they lost their state because the server was restarted. Additionally, specifying a server key lets load-balanced servers access the same information.
+
+Continuing our example from above... _(This should not be used on production servers; a key should be generated and provided via a configuration file.)_
+
+{% highlight fsharp %}
+let suaveCfg =
+  { defaultConfig with
+      serverKey = System.Text.Encoding.UTF8.GetBytes("12345678901234567890123456789012")
+    }
+
+[<EntryPoint>]
+let main argv = 
+  startWebServer suaveCfg app
+  0 
+{% endhighlight %}
+
+Cookie Serialization<br>_(of particular interest to .NET Core < netstandard2.0)_
+--------------------------------------------------------------------------------
+
+Suave uses the .NET Framework type `BinaryFormatter` to serialize the `Map<string, obj>` containing the session state; this is the default. However, the `BinaryFormatter` was removed in the .NET Core API, and the `DataContractJsonSerializer` does not recognize the `Map<string, obj>` type. One option is to utilize JSON.NET to serialize this object. To use that, ensure you've added the `Newtonsoft.Json` NuGet package to your project, then put the following code somewhere before the `suaveCfg` definition in the example above.
+
+{% highlight fsharp %}
+/// alias
+let utf8 = System.Text.Encoding.UTF8
+
+type JsonNetCookieSerialiser () =
+  interface CookieSerialiser with
+    member x.serialise m =
+      utf8.GetBytes (JsonConvert.SerializeObject m)
+    member x.deserialise m =
+      JsonConvert.DeserializeObject<Map<string, obj>> (utf8.GetString m)
+{% endhighlight %}
+
+Then, modify the configuration to use that serializer.
+
+{% highlight fsharp %}
+/// Again, do not make this your server key in production...
+let suaveCfg =
+  { defaultConfig with
+      serverKey = System.Text.Encoding.UTF8.GetBytes("12345678901234567890123456789012")
+      cookieSerialiser = new JsonNetCookieSerialiser()
+    }
+{% endhighlight %}
+
+State among WebParts
+--------------------
+
+Within the `Writers` module, Suave provides the functions `setUserData` and `unsetUserData` for adding items to the context's `userState` property.  The example below could be used to accrue a list of messags to be displayed to the user.
+
+{% highlight fsharp %}
+/// string -> WebPart
+let addUserMessage message =
+  context (fun ctx ->
+    let existing =
+      match ctx.userState.ContainsKey "messages" with
+      | true -> ctx.userState.["messages"] :?> string list
```

Comment:
Extract to `let read context key: 'value = context.userState |> Map.tryFind key |> Option.map (fun x -> x :?> 'value)` function? Then curry in example; `let read = read ctx` and use `read "messages"` which is then automatically typed and downcast to the right type.